sdk/data/azcosmos/cosmos_global_endpoint_manager.go

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azcosmos
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	azruntime "github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+)
+
+const defaultUnavailableLocationRefreshInterval = 5 * time.Minute
+
+type globalEndpointManager struct {
+	client              *Client
+	preferredLocations  []string
+	locationCache       *locationCache
+	refreshTimeInterval time.Duration
+}
+
+func newGlobalEndpointManager(client *Client, preferredLocations []string, refreshTimeInterval time.Duration) (globalEndpointManager, error) {
+	endpoint, err := url.Parse(client.endpoint)
+	if err != nil {
+		return globalEndpointManager{}, err
+	}
+
+	if refreshTimeInterval == 0 {
+		refreshTimeInterval = defaultUnavailableLocationRefreshInterval
+	}
+
+	gem := globalEndpointManager{
+		client:              client,
+		preferredLocations:  preferredLocations,
+		locationCache:       newLocationCache(preferredLocations, *endpoint),
+		refreshTimeInterval: refreshTimeInterval}
+
+	return gem, nil
+}
+
+func (gem *globalEndpointManager) GetWriteEndpoints() ([]url.URL, error) {
+	return gem.locationCache.writeEndpoints()
+}
+
+func (gem *globalEndpointManager) GetReadEndpoints() ([]url.URL, error) {
+	return gem.locationCache.readEndpoints()
+}
+
+func (gem *globalEndpointManager) MarkEndpointUnavailableForWrite(endpoint url.URL) error {
+	return gem.locationCache.markEndpointUnavailableForWrite(endpoint)
+}
+
+func (gem *globalEndpointManager) MarkEndpointUnavailableForRead(endpoint url.URL) error {
+	return gem.locationCache.markEndpointUnavailableForRead(endpoint)
+}
+
+func (gem *globalEndpointManager) GetEndpointLocation(endpoint url.URL) string {
+	return gem.locationCache.getLocation(endpoint)
+}
+
+func (gem *globalEndpointManager) CanUseMultipleWriteLocations() bool {
+	return gem.locationCache.canUseMultipleWriteLocs()
+}
+
+func (gem *globalEndpointManager) IsEndpointUnavailable(endpoint url.URL, ops requestedOperations) bool {
+	return gem.locationCache.isEndpointUnavailable(endpoint, ops)
+}
+
+func (gem *globalEndpointManager) RefreshStaleEndpoints() {
+	gem.locationCache.refreshStaleEndpoints()
+}
+
+func (gem *globalEndpointManager) Update() error {
```

Comment:
Flow a context in:

```suggestion
func (gem *globalEndpointManager) Update(ctx context.Context) error {
```

Then pass it to the `GetAccountProperties()` instead of using `context.Background()` inside